src/ModelGradient.cpp

```diff
@@ -151,11 +151,107 @@ void LRSparseGradient::serialize(void* mem) const {
   }
 }
 
+/**
+ * Given a pointer to memory, method will serialize gradient into 'parts' number
+ * of parts
+ * Method returns vector of size parts containing tuples of <a, b> such that for
+ * the ith tuple mem + a
+ * is the start of the ith serialized gradient and mem + a + b is the end of
+ * that serialized gradient
+ *
+ */
+
+std::array<std::tuple<int, int>, MAX_NUM_PS> LRSparseGradient::shard_serialize(
+    void* mem,
+    uint32_t parts) const {
+  std::array<int, MAX_NUM_PS> starts;  // First will store how many tuples per
+                                       // shard, ie starts[i] is the number
+  starts.fill(0);  // of gradient weights going to server number i
+  std::array<std::tuple<int, int>, MAX_NUM_PS>
+      starts_out;  // starts_out[i] stores <starting_offset(bytes), size(bytes)>
+                   // of where the ith gradient
+                   // serialize starts and its size relative to mem.
+  // TODO: Change to murmurhash
+  std::hash<uint32_t> hash;  // hash function
+
+  // Perform count
+  for (const auto& w : weights) {
+    starts[hash(w.first) % parts]++;
+  }
+
+  // starts[i] now will store how many weights into mem does the ith gradient
+  // serialzation start.
+
+  int count =
+      starts[0];  // how many weights belong in the current shard serialization
+  int count_next =
+      starts[1];  // how many weights belong in the next shard serialization
+
+  starts[0] = 0;      // First shard serialization starts at 0
```

Comment:
I think you overcomplicate this code here and in the next loop. It seems to me you only need two things: one array with the number of weights for each part, a cumulative sum of the number of weights for the parts. Same thing in the MF code later below. What do you think?

Reply:
I think what you described is what we're doing, but its not the most clear. 

In the first for loop I count the number of weights per part. 

In the 2nd for loop I turn the starts array to be a cumulative count but cumulative in the number of bytes. This for loop also rights in the gradient version and and count. 

Last for loop populates the gradient with values from the cumulative.

Reply:
I looked into this again and I still think this is too complicated.

What you are basically doing here is, given a vector of bucket counts (computed in the first loop of this function):

Example:
```
{1,2,3,4}
```

You compute a vector of positions, first position is 0, and the others are the cumulative sum. The result (which, you put in the vector ```starts```) is:

```
0, 1, 3, 6, 10
```

Your code (seems to me) is overly complex because you decide to overwrite starts, and then you overcomplicate on the count, and count_next vars.

Reply:
The simplest way would be to create a new std::array (same size as starts). No need to overwrite starts this way. Also, no need to fill the new vector with 0s since we are going to build it.

This new array (let's call it positions) is your new vector of positions. Then you do:

```
positions[0] = 0
for (int i = 0; i < parts; i++) {
    count = starts[i]; // or just use starts[i] directly in the omitted code
    if (i != (parts - 1) && i != 0) {
      positions[i] = positions[i-1] + starts[i];
    }
    ....
```

Please correct me If I'm wrong.